### DISCUSSION TITLE
Adding custom advancement frames

### DISCUSSION THREAD

**Discussion by Desynq** [FabricMC/fabric-api#2346] [Mod Dev Support] [answered]
Looking at the source code for advancement frames, it seems to be an enum. I'm wondering if there's any way I can add my own custom advancement frames since I'm guessing you can't extend it like you can when making custom items.

My best guess is that I'd have to do mixin injection or possibly use Fabric-ASM. If so, can anyone give me an example of how I would do this please, since I'm still relatively inexperienced in mixin.

**Comment by warjort** [comment, marked as answer]
Yes you will need to use fabric-asm to modify an enum, mixin doesn't support it.
The docs for fabric-asm point to this example: https://github.com/Chocohead/Fabric-ASM/blob/master/example/src/com/chocohead/mm/testing/EarlyRiser.java#L22-L29

But, you will be "creating a rod for your own back" doing this kind of stuff.

Consider what will happen in this mod's code with your new enum value:
https://github.com/MehVahdJukaar/AdvancementFrames/blob/6ace47748fa4410d057e79bb3d5aea6341029a66/src/main/java/net/mehvahdjukaar/advframes/client/AdvancementFrameBlockTileRenderer.java#L65
The answer is it will crash! This is why mixin doesn't support this stuff.

In this case, it is a forge mod, but who knows what other mods are out there doing similar things?
To avoid these mod incompatibilities you will need to creates patches (mixins to these other mods) in case they are installed and need to know how to handle your new values.

**Reply by Desynq** [reply]
ok, thanks. How can get the class id for a Minecraft class? Similar to how your listed example references `net.minecraft.class_2582`

Also, is it possible for you to give me a brief example of how to implement this example into a development environment? I could do fine with just knowing the maven, groupId, artifactId, and recommended version, thanks. Unless it has to implemented differently.

**Reply by warjort** [reply]
You need to lookup the yarn mappings to get the intermediary name, e.g. https://github.com/FabricMC/yarn/blob/1.19/mappings/net/minecraft/advancement/AdvancementFrame.mapping

I have never used fabric-asm, if you can't tell from my answer above, I think it is very hacky and more trouble than it is worth. :-)
You can find many examples of its usage github:
https://github.com/search?q=fabric-asm&type=code

**Reply by Desynq** [reply]
ok thanks.
